src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,372 @@
+    /// wildcard pointers are used to access this location. What we do know is that `borrows` are at
+    /// the top of the stack, and below it are arbitrarily many items whose `tag` is strictly less
+    /// than `id`.
+    /// When the bottom is unknown, `borrows` always has a `SharedReadOnly` or `Unique` at the bottom;
+    /// we never have the unknown-to-known boundary in an SRW group.
+    unknown_bottom: Option<SbTag>,
+
+    /// A small LRU cache of searches of the borrow stack. This only caches accesses of `Tagged`,
+    /// because those are unique in the stack.
+    #[cfg(feature = "stack-cache")]
+    cache: StackCache,
+    /// On a read, we need to disable all `Unique` above the granting item. We can avoid most of
+    /// this scan by keeping track of the region of the borrow stack that may contain `Unique`s.
+    #[cfg(feature = "stack-cache")]
+    unique_range: Range<usize>,
+}
+
+/// A very small cache of searches of the borrow stack
+/// This maps tags to locations in the borrow stack. Any use of this still needs to do a
+/// probably-cold random access into the borrow stack to figure out what `Permission` an
+/// `SbTag` grants. We could avoid this by also storing the `Permission` in the cache, but
+/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
+#[cfg(feature = "stack-cache")]
+#[derive(Clone, Debug)]
+struct StackCache {
+    tags: [SbTag; CACHE_LEN], // Hot in find_granting
+    idx: [usize; CACHE_LEN],  // Hot in grant
+}
+
+#[cfg(feature = "stack-cache")]
+impl StackCache {
+    /// When a tag is used, we call this function to add or refresh it in the cache.
+    ///
+    /// We use the position in the cache to represent how recently a tag was used; the first position
+    /// is the most recently used tag. So an add shifts every element towards the end, and inserts
+    /// the new element at the start. We lose the last element.
+    /// This strategy is effective at keeping the most-accessed tags in the cache, but it costs a
+    /// linear shift across the entire cache when we add a new tag.
+    fn add(&mut self, idx: usize, tag: SbTag) {
+        self.tags.copy_within(0..CACHE_LEN - 1, 1);
+        self.tags[0] = tag;
+        self.idx.copy_within(0..CACHE_LEN - 1, 1);
+        self.idx[0] = idx;
+    }
+}
+
+impl PartialEq for Stack {
+    fn eq(&self, other: &Self) -> bool {
+        // All the semantics of Stack are in self.borrows, everything else is caching
+        self.borrows == other.borrows
+    }
+}
+
+impl Eq for Stack {}
+
+impl<'tcx> Stack {
+    /// Panics if any of the caching mechanisms have broken,
+    /// - The StackCache indices don't refer to the parallel tags,
+    /// - There are no Unique tags outside of first_unique..last_unique
+    #[cfg(feature = "expensive-debug-assertions")]
+    fn verify_cache_consistency(&self) {
+        if self.borrows.len() > CACHE_LEN {
```

Comment:
```suggestion
        // Only a full cache needs to be valid. Also see the comments in find_granting_cache
        // and set_unknown_bottom.
        if self.borrows.len() > CACHE_LEN {
```
The one thing I am surprised by is that this is not `>=`. But it is consistent with `find_granting_cache` so seems fine.

Reply:
Nah, I'll bump it up to `>=`. The `==` case ought to be valid.